### ISSUE TITLE
[bug] onResolve is ignored.

### ISSUE THREAD

**Issue by jodosha** [evanw/esbuild#3005]
TL;DR I want to achieve the following path resolution:

  * `app/assets/javascripts/index.js -> public/assets/index-abc123.js`
  * `slices/admin/javascripts/index.js -> public/assets/admin/index-def456.js`

This is NOT the default esbuild behavior, so I created a [plugin](https://github.com/hanami/esbuild).
The plugin uses `onResolve` callback, which seems to be invoked, but the returning `path` is ignored.

---

esbuild’s plugins can intercept importing paths and decide the resolution path.
I need this to control the destination path (see the mapping above).

The callback to use is `onResolve` ([doc](https://esbuild.github.io/plugins/#on-resolve)):
The callback can customize how esbuild does path resolution. For example, it can intercept import paths and redirect them somewhere else.

The example explains that I should set the returning value of the callback as `{ path: "…" }`.

> `path`: Set this to a non-empty string to resolve the import to a specific path. If this is set, no more on-resolve callbacks will be run for this import path in this module. If this is not set, esbuild will continue to run on-resolve callbacks that were registered after the current one. Then, if the path still isn’t resolved, esbuild will default to resolving the path relative to the resolve directory of the current module.

I do what is expected: [return the absolute destination path](https://github.com/hanami/esbuild/blob/d6c6597cab1c3f105c8262173bff8699e41ef90d/src/hanami-esbuild.ts#L41).

👉 **BUT esbuild fails the build, saying that the destination path doesn’t exist.** 👈 

```shell
⚡ tree tmp
tmp
├── app # OK: ORIGINAL SOURCE
│   └── assets
│       ├── images
│       └── javascripts
│           └── index.js
├── public
│   └── assets # WRONG: EMPTY DIR
│       └── admin  # WRONG: EMPTY DIR
└── slices # OK: ORIGINAL SOURCE
    └── admin
        └── assets
            └── javascripts
                └── index.js

11 directories, 2 files
```

```shell
⚡ npm test

> hanami-esbuild@1.0.0 test
> jest

 FAIL  test/hanami-esbuild.test.ts
  hanamiEsbuild
    ✕ copies assets from app/assets to public/assets and generates a manifest file (27 ms)

  ● hanamiEsbuild › copies assets from app/assets to public/assets and generates a manifest file

    Build failed with 2 errors:
    error: Could not read from file: /Users/jodosha/Code/hanami/hanami-esbuild/tmp/public/assets/index.js
    error: Could not read from file: /Users/jodosha/Code/hanami/hanami-esbuild/tmp/public/assets/admin/index.js

      at failureErrorWithLog (node_modules/esbuild/lib/main.js:1636:15)
      at node_modules/esbuild/lib/main.js:1048:25
      at runOnEndCallbacks (node_modules/esbuild/lib/main.js:1471:45)
      at buildResponseToResult (node_modules/esbuild/lib/main.js:1046:7)
      at node_modules/esbuild/lib/main.js:1075:16
      at responseCallbacks.<computed> (node_modules/esbuild/lib/main.js:697:9)
      at handleIncomingPacket (node_modules/esbuild/lib/main.js:752:9)
      at Socket.readFromStdout (node_modules/esbuild/lib/main.js:673:7)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        1.634 s, estimated 2 s
Ran all test suites.
```

[[ci](https://github.com/hanami/esbuild/actions/runs/4466654628/jobs/7845113723)]

---

This is the esbuild [configuration used during the unit tests](https://github.com/hanami/esbuild/blob/d6c6597cab1c3f105c8262173bff8699e41ef90d/test/hanami-esbuild.test.ts#L45-L55):

```ts
    await esbuild.build({
      entryPoints: [entryPoint1, entryPoint2],
      bundle: true,
      outdir: outDir,
      loader: loader,
      logLevel: "silent",
      minify: true,
      sourcemap: true,
      entryNames: "[name]-[hash]",
      plugins: [hanamiEsbuild()],
    });
```

```ts
entryPoint1 /Users/jodosha/Code/hanami/hanami-esbuild/tmp/app/assets/javascripts/index.js
entryPoint2 /Users/jodosha/Code/hanami/hanami-esbuild/tmp/slices/admin/assets/javascripts/index.js
outDir /Users/jodosha/Code/hanami/hanami-esbuild/tmp/public/assets
loader {}
```

👉 I do change the current CWD [before](https://github.com/hanami/esbuild/blob/d6c6597cab1c3f105c8262173bff8699e41ef90d/test/hanami-esbuild.test.ts#L20) and [after](https://github.com/hanami/esbuild/blob/d6c6597cab1c3f105c8262173bff8699e41ef90d/test/hanami-esbuild.test.ts#L25) each test to simulate a command run at the root of the app. In my case, the app's root is `tmp/`.

---

If I comment out the `onResolve` callbacks, esbuild falls back to the default behavior, and it adds the hash to the file path name:

```shell
⚡ tree tmp
tmp
├── app
│   └── assets
│       ├── images
│       └── javascripts
│           └── index.js
├── public
│   └── assets
│       ├── index-AJQHFXBY.js # OK
│       ├── index-AJQHFXBY.js.map # OK
│       ├── index-M3OZRUS7.js  # NOT OK, I want it under public/assets/admin directory
│       └── index-M3OZRUS7.js.map   # NOT OK, I want it under public/assets/admin directory
└── slices
    └── admin
        └── assets
            └── javascripts
                └── index.js

10 directories, 6 files
```

**Comment by KonnorRogers** [comment]
Related issues:

#553 
#2985
